### PR TITLE
[10.0] mig pos default empty image

### PR DESCRIPTION
**Pull request by legalsylvain**
Port and improve pos_default_empty_image from v8.

This PR is the sum of : 
- c60331f 3b4a4ff 62575ce 6dea09e 3293d65 b2adf04 : the cherry pick from V8.0 module. (https://github.com/OCA/pos/pull/82)
- b5f67f0 : the port of the module made by @invitu (https://github.com/OCA/pos/pull/154)
- bc0906a : improvement made by @invitu (https://github.com/OCA/pos/pull/156)
- 5bd182d : improvement and fixes.

thanks for your review. CC : 
@hparfr, @invitu, @bealdav, @lasley, @pedrobaeza.

### CODE REVIEW

File: pos_default_empty_image/README.rst [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+POS Default Empty Image
+=======================
+
+This module extends the functionality of point of sale to make PoS load faster
+and to improve products display.
+
+Point Of Sale Load faster
+-------------------------
+
+In the point of sale, trying to load known inexistant images is a waste of time.
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: rephrase to remove "waste of time" (eg: "loading product without image spend unnecessary time and resources.")

File: pos_default_empty_image/README.rst [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+When you have 8000 products in your Point of Sale and most of them 
+don't have images, you are happy to save thousands of useless requests:
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: remove "you are happy". eg: removing thousands of useless requests is welcome:

File: pos_default_empty_image/README.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+the POS load way faster.
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: the POS loads faster that way

File: pos_default_empty_image/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# © <2015> <Akretion, GRAP, OCA>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    'name': 'POS Default empty image',
+    'version': '10.0.0.1.0',
```
Comment (elicoidal) [inline pos_default_empty_image/__manifest__.py]: 10.0.1.0.0

File: pos_default_empty_image/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    'category': 'Point Of Sale',
+    'summary': 'Optimise load time for products with no image',
```
Comment (elicoidal) [inline pos_default_empty_image/__manifest__.py]: loading / without

**Review by elicoidal — CHANGES_REQUESTED** [review]
some wording

File: pos_default_empty_image/static/src/js/widgets.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        // Change product display if product has no image;
+        render_product: function(product){
+            if (product.has_image){
+                return this._super(product);
+            }
+            else {
+                var cached = this.product_cache.get_node(product.id);
+                if(!cached){
+                    var image_url = this.get_product_image_url(product);
```
Comment (hparfr) [inline pos_default_empty_image/static/src/js/widgets.js]: useless ?

**Review by hparfr — CHANGES_REQUESTED** [review]

File: pos_default_empty_image/static/src/js/widgets.js (hunk 6 above)
Comment (legalsylvain) [reply pos_default_empty_image/static/src/js/widgets.js]: indeed !

**Comment by legalsylvain** [comment]
@hpar, @elicoidal : Thanks for your review.
Changes done.

**Comment by hparfr** [comment]
👍

File: pos_default_empty_image/README.rst [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+================================================
+
+This module extends the functionality of point of sale to make PoS load faster
+and to improve products display.
+
+Point Of Sale Load faster
+-------------------------
+
+In the point of sale, loading product without image spend unnecessary time
+and resources
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: missing period

File: pos_default_empty_image/models/product_product.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# © <2015> <Akretion, GRAP, OCA>
```
Comment (elicoidal) [inline pos_default_empty_image/models/product_product.py]: remove <>

**Comment by legalsylvain** [comment]
done again.

File: pos_default_empty_image/README.rst [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+================================================
+Optimize loading time for products without image
+================================================
+
+This module extends the functionality of point of sale to make PoS load faster
+and to improve products display.
+
+Point Of Sale Load faster
+-------------------------
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: Sorry I just saw it but the heading level needs to be consistent with the following ones. It should be "===="
Actually you should probably add an intermediate level 1 title "introduction"

Comment (legalsylvain) [reply pos_default_empty_image/README.rst]: No problem ! Thanks a lot for your complete review !

**Comment by rafaelbn** [comment]
Hi @legalsylvain I'm testing this module but when POS loads all the images are there. I expected to see all products in POS ***without*** images. Isn't it?

**Comment by JGarcia-Panach** [comment]
Hi @legalsylvain, I've tested in runbot and I really like the new display of products with no image.

Just a minor detail, if the name is too long the display is not so good. Maybe the name should be truncated?

Example
image

On the other hand, I have the same question as @rafaelbn about seeing all products with images.

Thanks!

**Comment by legalsylvain** [comment]
> Hi @legalsylvain I'm testing this module but when POS loads all the images are there. I expected to see all products in POS without images. Isn't it?

Hi @rafaelbn and @JGarcia-Panach. No this is really not the objective of this module.
The aim of this module (since V8) is to avoid to load images when products have no images. (Please remove all images from your products and check your logs. You'll see a lot of request loading the **default image** for each products. There calls are really useless.
this module fixes that, removing that usell noices on the network, and improving loading time. it's like a bug fix of the Odoo Core PoS, and should be installed on all instance by default. 

My PoV is that removing images can be done, but in another module. like *pos_product_image_no_load*, or by an option on this module. But it is currently out of the scope of my PR. (just porting and cleaning this module).

> Just a minor detail, if the name is too long the display is not so good. Maybe the name should be truncated?

I didn't changed V8 behaviour. I'm not a CSS expert and i don't know how to do that but if you make a PR against mine, with the CSS changes, I'll review it and 'll merge it quickly. (or it can be done in a second time).
https://github.com/OCA/pos/pull/164/files#diff-3c016e2d49e157baaa1509c4cbb73953

@elicoidal : Is it now ok for you, or do you have others remarks ?

kind regards, and thanks for your review.

**Comment by lasley** [comment]
For the CSS - [`overflow-wrap: break-word`](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-wrap)

**Comment by rafaelbn** [comment]
👍

**Review by rafaelbn — APPROVED** [review]
Tested in runbot!

File: pos_default_empty_image/README.rst [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+With this module, the display of the product is changed, (Size of the name
+is increased for better visibility);
+
+  .. figure:: /pos_default_empty_image/static/description/pos_display_improved.png
+     :width: 800 px
+
+Technical information
+=====================
+
+Each time the pos instantiate a product, it will add this code for each product
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: PoS

File: pos_default_empty_image/README.rst [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (elicoidal) [inline pos_default_empty_image/README.rst]: https

File: pos_default_empty_image/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# © <2015> <Akretion, GRAP, OCA>
```
Comment (elicoidal) [inline pos_default_empty_image/__manifest__.py]: remove the <>

**Review by elicoidal — APPROVED** [review]
small details. thanks for the contribution

**Comment by rafaelbn** [comment]
Hi @legalsylvain , check minor comments is RTM!

**Review by hparfr — APPROVED** [review]

**Pull request merged**